### PR TITLE
Add support for `datetimeoffset` MsSql data type

### LINKED ISSUE #1358: DAB - Column type DateTimeOffset not yet supported
### Discussed in https://github.com/Azure/data-api-builder/discussions/1356

<div type='discussions-op-text'>

<sup>Originally posted by **nihasalangaden** March 24, 2023</sup>
Hi All,

I was trying this with an existing table , but getting the below error: any thoughts?

dbug: Azure.DataApiBuilder.Service.Services.ISqlMetadataProvider[0]
      Logging primary key information for entity: test.
dbug: Azure.DataApiBuilder.Service.Services.ISqlMetadataProvider[0]
      Primary key column name: application_id
      Primary key mapped name: application_id
      Type: Byte
      IsNullable: False
      IsAutoGenerated: False
fail: Azure.DataApiBuilder.Service.Startup[0]
      Unable to complete runtime initialization operations due to:
System.ArgumentException: Column type DateTimeOffset not yet supported.
         at Azure.DataApiBuilder.Service.Parsers.EdmModelBuilder.BuildEntityTypes(ISqlMetadataProvider sqlMetadataProvider)
         at Azure.DataApiBuilder.Service.Parsers.EdmModelBuilder.BuildModel(ISqlMetadataProvider sqlMetadataProvider)
         at Azure.DataApiBuilder.Service.Parsers.ODataParser.BuildModel(ISqlMetadataProvider sqlMetadataProvider)
         at Azure.DataApiBuilder.Service.Services.SqlMetadataProvider`3.InitODataParser()
         at Azure.DataApiBuilder.Service.Services.SqlMetadataProvider`3.InitializeAsync()
         at Azure.DataApiBuilder.Service.Startup.PerformOnConfigChangeAsync(IApplicationBuilder app)
info: Azure.DataApiBuilder.Service.Startup[0]
      Loading config file: dab-config.json
fail: Azure.DataApiBuilder.Service.Startup[0]
      Exiting the runtime engine...
**Table structure is as below:**
application_id	tinyint
application_cd	varchar
application_nm	varchar
create_ts	datetimeoffset
update_ts	datetimeoffset

Config file added like below:
"test": {
        "source": {
          "type": "table",
          "object": "dbo.application_ref",
          "key-fields":["application_id"]
        },
        "rest": {
          "methods": [ "get" ]
        },
        "graphql": {
          "operation": "query"
        },
        "permissions": [{
         "role": "anonymous",
          "actions": [ "*" ]
        }]
      }</div>

### PR DESCRIPTION
**Pull request by Aniruddh25**
## Why make this change?

- Closes #1358 
by considering this case in the EdmModelBuilder and SchemaConverter.

## What is this change?

- Add an additional case for `DateTimeOffset` while converting from `SystemType` to `EdmPrimitiveTypeKind`.
- `DateTimeOffset` is simply `DateTime` in HotChocolate.Types used in SchemaConverter.
- Add a new `DateTimeOffset` case to Param creation.

## How was this tested?

- Added new column with `datetimeoffset` in MsSql to test this. Also added columns for `datetime2`, `date`, `smalldatetime` to ensure those are already supported
- [X] Integration Tests
- [X] Unit Tests

### CODE REVIEW

**Review by ayush3797 — APPROVED** [review]
LGTM!

**Review by abhishekkumams — APPROVED** [review]
LGTM!

**Review by seantleonard — COMMENTED** [review]
Looks good so far, is there any way to add 1 test for e2e REST request with this new datatype?

**Comment by Aniruddh25** [comment]
> Looks good so far, is there any way to add 1 test for e2e REST request with this new datatype?

There is an existing issue to add new tests for REST - #662 for additional supported types

File: src/Service.GraphQLBuilder/Sql/SchemaConverter.cs [hunk 1]
```diff
@@ -243,6 +243,7 @@ public static string GetGraphQLTypeForColumnType(Type type)
                 "Decimal" => DECIMAL_TYPE,
                 "Boolean" => BOOLEAN_TYPE,
                 "DateTime" => DATETIME_TYPE,
+                "DateTimeOffset" => DATETIME_TYPE,
```
Comment (seantleonard) [inline src/Service.GraphQLBuilder/Sql/SchemaConverter.cs:246]: in my storedprocedure param valuetype pr i needed to change this to have a dedicated datatimeoffset_type to handle differences between DataTime and DateTimeOffset.

Comment (seantleonard) [reply src/Service.GraphQLBuilder/Sql/SchemaConverter.cs:246]: I didn't end up changing this.

**Pull request merged**